torch/csrc/jit/codegen/cuda/ir_base_nodes.h

```diff
@@ -391,6 +404,31 @@ class TORCH_CUDA_CU_API Val : public Statement {
   int evaluator_index_ = -1;
 };
 
+//! A Val object that stores a plain data. Note that this class is only intended
+//! to hold non-IR data, such as DataType, std::vector<int>, etc. Please don't
+//! use this class to hold IR nodes or their pointers.
+template <typename T>
+class TORCH_CUDA_CU_API PlainVal : public Val {
```

Comment:
So, this is intended to hold non-IR data, but then why is it a subclass of `Val`?

Reply:
So, this is used to hold IR node attributes. Does it need to be a subclass of `Val`?

Reply:
Because `attributes_` are vectors of `Val*` and I want to push plain values into `attributes_`. Maybe I should change `attributes_` to a vector of `Statement*`? Currently, lots of kernel IR has expr members, and I am storing them as something like `PlainVal<Statement*>`; maybe making `attributes_` a vector of `Statement*` would make it easier because I can directly store it.

Reply:
Ah, I see. Yeah, we want to store proper IR-node values as well. Can it be `std:vector<std::any>`?

Reply:
Not today. `std::any` is C++17 but PyTorch now is still using C++14

Reply:
OK, but my question still holds. Can we just have something like:

```
class ExprAttribute {
};

template <typename T>
class ExprAttribute: AbstractExprAttribute {
   T data_;
   ...
};
```

Then in `Expr`, we could have `std::vector<AbstractExprAttribute> attributes_`.

Reply:
I think it can only be `std::vector<std::shared_ptr<AbstractExprAttribute>> attributes_`?

Reply:
Ah, right. Can it be a vector of `unique_ptr`? 

I just feel odd to have arbitrary attributes, like a `DataType` and a bool flag, as a `Val`. Please convince me if you really want to use `PlainVal`.

Reply:
I believe the biggest advantage of making it a `Val` is that, `IrCloner` will automatically help you do deep copy when necessary. For example, if I have an attribute `bool`, and I clone the expr, then I will also have a new `PlainVal<bool>` object for the new expr. If later, I changed the new expr's bool attribute, the old one's bool attribute is not changed. But if we manage this by hand, we could make mistakes that, when cloning an expr, the old and new expr are pointing to the same object for its attribute, and if we change one, the other is also changed.

So to answer your question:
> Can it be a vector of `unique_ptr`?

Yes. Actually, it SHOULD be a `unique_ptr` instead of a `shared_ptr`, for the reason I described above.

But without using `PlainVal`, we will have to manually do a deep copy of the `std::vector<std::unique_ptr<AbstractExprAttribute>> attributes_`. It's a little inconvenient, but should be more efficient.

I don't have a strong preference about which one we should choose.

Reply:
Oh, wait, no. It is not "a little inconvenient", but super convenient. Without knowing `T`, we can not do a deep copy.

Reply:
OK, that makes sense. (I assume you meant "super inconvenient")

Let's use `PlainVal`, but I'd prefer a name like `Attribute`, which I think a little more clearly implies what it's supposed to be used for.

Reply:
Renamed